### ISSUE TITLE
Real-time annotation of EEG data

### ISSUE THREAD

**Issue by Remi-Gau** [brainhackorg/global2021#196]
Added as an issue for book keeping

Source:
https://brainhack.ch/past_events/global2021/index.html#bsl-annotations
https://github.com/brainhack-ch/bsl

Add a real-time annotation system compatible with MNE to mark bad segments of EEG data on BSL's real-time viewer using pyqtgraph and pyopengl 

Leaders:
[Mathieu Scheltienne](https://brainhack.ch/past_events/global2021/index.html#mathieu_scheltienne)

EEG data is noisy by nature, and MNE has a very nice feature called [annotations](https://mne.tools/stable/auto_tutorials/raw/30_annotate_raw.html#sphx-glr-auto-tutorials-raw-30-annotate-raw-py) to mark bad segments of data. Those segments are then rejected internally by functions accepting the argument reject_by_annotation.
Until recently, interactive annotation on the MNE browser was very painful as the default browser (up to version 1.0) uses an old matplotlib backend... as slow as it gets. But, with version 0.24, a [new backend](https://github.com/mne-tools/mne-qt-browser) has been introduced using pyqtgraph and pyopengl to render on the GPU. And it is amazing!
But even with this new backend, annotating raw data takes a lot of time, especially on large scale study.
This is where this project comes in. During a typical EEG session, someone will be monitoring the signal. In itself, this is a lightweight task, that could be coupled with a system to annotate in real-time the monitored signal. No more annotation in a later post-processing stage!

The FCBG platform has develop a library called [NeuroDecode](https://github.com/fcbg-hnp/NeuroDecode) which has a build-in real-time viewer. This library uses LSL to stream EEG data, and pyqtgraph (rings a bell?) for the viewer. All the data streaming/acquisition/viewing capabilities have been moved and improved in [BSL](https://bsl-tools.github.io/).

Basically, this project doesn't start from scratch! You can find [here](https://bsl-tools.github.io/_images/stream_viewer.gif) an example of the real-time viewer that will be modified for this project. The viewer has 2 backends: pyqtgraph and vispy. The second is incomplete, but runs efficiently on GPU. However, it is a bit obsolete as new version of pyqtgraph are now compatible with pyopengl... to run efficiently on GPU.

The objective is to add pyopengl, improve the GUI/add an annotation section, and add click and drag annotation on the backend window.
